website/addons/googledrive/model.py

```diff
@@ -111,218 +101,183 @@ def get_or_create(cls, node, path):
+            self.account.save()
 
     def _needs_refresh(self):
-        if self.expires_at is None:
+        if self.account.expires_at is None:
             return False
-        return (self.expires_at - datetime.utcnow()).total_seconds() < drive_settings.REFRESH_TIME
+        return (self.account.expires_at - datetime.utcnow()).total_seconds() < drive_settings.REFRESH_TIME
 
 
-class GoogleDriveUserSettings(AddonUserSettingsBase):
-    """Stores user-specific information, including the Oauth access
-    token.
-    """
-    oauth_settings = fields.ForeignField(
-        'googledriveoauthsettings', backref='accessed'
-    )
+class GoogleDriveUserSettings(StorageAddonBase, AddonOAuthUserSettingsBase):
+    oauth_provider = GoogleDriveProvider
+    oauth_grants = fields.DictionaryField()
+    serializer = GoogleDriveSerializer
 
-    @property
-    def user_id(self):
-        if self.oauth_settings:
-            return self.oauth_settings.user_id
-        return None
 
-    @user_id.setter
-    def user_id(self, val):
-        self.oauth_settings.user_id = val
+class GoogleDriveNodeSettings(StorageAddonBase, AddonOAuthNodeSettingsBase):
+    oauth_provider = GoogleDriveProvider
 
-    @property
-    def username(self):
-        if self.oauth_settings:
-            return self.oauth_settings.username
-        return None
+    drive_folder_id = fields.StringField(default=None)
+    drive_folder_name = fields.StringField(default=None)
+    folder_path = fields.StringField()
+    serializer = GoogleDriveSerializer
 
-    @username.setter
-    def username(self, val):
-        self.oauth_settings.username = val
+    _api = None
 
     @property
-    def access_token(self):
-        if self.oauth_settings:
-            return self.oauth_settings.access_token
-        return None
-
-    @access_token.setter
-    def access_token(self, val):
-        self.oauth_settings.access_token = val
+    def api(self):
+        """Authenticated ExternalProvider instance"""
+        if self._api is None:
+            self._api = GoogleDriveProvider()
+            self._api.account = self.external_account
+        return self._api
 
     @property
-    def refresh_token(self):
-        if self.oauth_settings:
-            return self.oauth_settings.refresh_token
-        return None
-
-    @refresh_token.setter
-    def refresh_token(self, val):
-        self.oauth_settings.refresh_token = val
-
-    @property
-    def expires_at(self):
-        if self.oauth_settings:
-            return self.oauth_settings.expires_at
-        return None
+    def complete(self):
+        return bool(self.has_auth and self.user_settings.verify_oauth_access(
+            node=self.owner,
+            external_account=self.external_account,
+            metadata={'folder': self.drive_folder_id}
+        ))
 
-    @expires_at.setter
-    def expires_at(self, val):
-        self.oauth_settings.expires_at = val
+    def set_folder(self, folder, auth, add_log=True):
```

Comment:
This method does not seem to be an override, but the call signature specifies two parameters that are not used in the function. Why?